### PR TITLE
add logic to show introduction paragraph is presented differently in the editing view

### PR DESCRIPTION
**Pull request by rajparad**
**Phabricator:** https://phabricator.wikimedia.org/T212318

### Notes
* This PR covers logic to show alert when clicking on edit introduction. Alert has been shown only once, when clicked "Yes". 
* This is my first PR to WIKI. Need suggestions to confirm the localize string implementation is correct or not. also ticket only mentions to have something so the user knows the order is different. based on that I have used alert to show only once.

### Test Steps
1. Open any article.
2. Click on pencil in right.
3. Click on "Edit introduction"
4. the new alert will appear.
5. Click "Yes" to proceed, click "cancel" and the repeat the flow again to show Alert.
6. If clicked "Yes" and want to see Alert again, uninstall app and re-install again.

### Screenshots/Videos
- Eye (Alexander McQueen
- Eye (Alexander McQueen
- The ordering of elements will

### CODE REVIEW

**Comment by tonisevener** [comment]
Hi @rajparad! Apologies for the review delay on this one. This all looks good from a code perspective - nice work! I will ask our designers to review further in case they have any additional behavior or copy tweaks before merging. Thanks!

**Comment by rajparad** [comment]
> Hi @rajparad! Apologies for the review delay on this one. This all looks good from a code perspective - nice work! I will ask our designers to review further in case they have any additional behavior or copy tweaks before merging. Thanks!

No worries, Thanks for the reply!! let me know if there is any feedback from designers :)

**Comment by rajparad** [comment]
Hello @tonisevener, There was a feedback from design to show alert. there are 2 ways.

Let me know what is your preference from code perspective. I can update accordingly :).

Design feedback is here: https://phabricator.wikimedia.org/T212318

**Comment by tonisevener** [comment]
> Hello @tonisevener, There was a feedback from design to show alert. there are 2 ways.
> 
> Let me know what is your preference from code perspective. I can update accordingly :).
> 
> Design feedback is here: https://phabricator.wikimedia.org/T212318

Hi @rajparad - apologies for the delay. I think we can get pretty close to option #1. You'll want to put in in the correct places in `PageEditorViewController`. We have a toast component that can be triggered like this:

```
WMFAlertManager.sharedInstance.showWarningAlert("Text here", sticky: false, dismissPreviousAlerts: true)
```

The design won't match up exactly, but that is okay.

screenshot

You can trigger within [loadContent](https://github.com/wikimedia/wikipedia-ios/blob/main/Wikipedia/Code/PageEditorViewController.swift#L202) once all the content is loaded in this method, if no issues have cropped up (i.e. no blocked error, no edit notice is displayed, etc.).

If an edit notice modal is automatically presented, you also want to try triggering it after that edit notice modal is dismissed. Usually the featured article of the day has an edit notice, if you want to test that and see how it behaves.

Hope this helps. Let us know if you have any other questions, thanks!

**Comment by rajparad** [comment]
Hello @tonisevener, Thanks for the feedback. I have updated the code with warning. 

image

Here is the link: https://github.com/wikimedia/wikipedia-ios/pull/4790/files

Only thing I am curious is it's now only work with English. I am not how to work with localization in project. Let me know if it's fine or not.

File: Wikipedia/Code/PageEditorViewController.swift
Comment (tonisevener) [inline Wikipedia/Code/PageEditorViewController.swift:251]: @rajparad I would like to only show this if other messages aren't triggered above, to prevent competing messages. Can you add an additional conditional here to only show the `editArticleWarning` if lines 221, 223, 225, and 233 were not hit (i.e. no blocked message, protected page message, other errors, or edit notices)? Just keeping track of a simple flag will suffice. Thanks!

Comment (rajparad) [reply Wikipedia/Code/PageEditorViewController.swift:251]: sure :). I will update and let you know

File: Wikipedia/Code/EditorViewController.swift
Comment (tonisevener) [inline Wikipedia/Code/EditorViewController.swift:255]: There is an additional change to the business requirements, stated here - https://phabricator.wikimedia.org/T212318#9928399.

To accomplish each, what I would do is:

> Ensure the notice only shows if someone has chosen to "Edit Introduction" from the first edit button, or "Edit Source" from the overflow menu. Currently I see the notice when I start editing a section, and I understand that the difference in ordering should not be an issue with section edits.

For this you can try adding a simple check against this view controller's `sectionID` property. If the sectionID == nil or 0, that should target from the paths in the new requirement. You can add it as a conditional here.

> The notice disappeared very quickly and I didn't have time to read all of it - is that our normal length for this type of alerts? or could it be extended?

You can update the `WMFAlertManager`'s [showWarningAlert](https://github.com/wikimedia/wikipedia-ios/blob/main/Wikipedia/Code/WMFAlertManager.swift#L59) method to allow passing in a custom duration. Then pass in 5 seconds from here. I will kick off another build and will comment on the task for our further review.

**Review by tonisevener — CHANGES_REQUESTED** [review]
@rajparad Thanks! A few more changes needed. Also our project changed a bit with some recent merges (a git submodule was deleted, and the name of `PageEditorViewController` changed to `EditorViewController`. Hopefully this doesn't cause issues when you pull in the latest from `main`.

File: Wikipedia/Code/EditorViewController.swift
Comment (rajparad) [reply Wikipedia/Code/EditorViewController.swift]: make sense :), I have updated the code for this

File: Wikipedia/Code/EditorViewController.swift
Comment (rajparad) [reply Wikipedia/Code/EditorViewController.swift]: Sure I will take a look, and let you know once I am done with changes :)

**Comment by rajparad** [comment]
> @rajparad Thanks! A few more changes needed. Also our project changed a bit with some recent merges (a git submodule was deleted, and the name of `PageEditorViewController` changed to `EditorViewController`. Hopefully this doesn't cause issues when you pull in the latest from `main`.

There were no major conflicts apart from having conflict in Localizable string. I still have that and when I try to resolve I don't see any changes coming from my local machine or from Above resolve conflict option in github.

**Comment by tonisevener** [comment]
@rajparad Yeah, that could be because git treats the files in `iOS Native Localizations` as binary. Generally when I run into strings conflicts, locally I choose the changes from `main`. Then the next time I build, our build script will automatically regenerate the strings file and cause git to show changes. Then I commit the new changes. Let me know if that works for you, thanks!

File: Wikipedia/Code/EditorViewController.swift
Comment (rajparad) [reply Wikipedia/Code/EditorViewController.swift]: Hello @tonisevener, Sorry for the delay!

I have pushed changes for sectionId and also fixed conflicts.

For `WMFAlertManager`'s `showWarningAlert` method change, I did modified the function but due to `@objc` notation it is not allowing me to pass extra param. 
image

I also tried to create via `WMFAppViewController.m` by creating an `@interface`, but no luck. Sorry, I am not really familiar with Obj-c.

Though One thing we can do is, use `true` for `showWarningAlert`'s `sticky` param to increase the duration to 2 seconds.

Let me know what you think!

Comment (tonisevener) [reply Wikipedia/Code/EditorViewController.swift]: @rajparad No problem, thanks for your patience on this. Sometimes we have to fall back to an NSNumber in order to allow optionality with Objective-C.  I pushed a commit doing this and passed in a duration of 5.0.

`sticky=true` means the banner would not automatically dismiss at all. This could have been fine but I noticed that it seems our warning messages do not switch themes (and theme controls are very easily accessed in the editor). Rather than go down the path of fixing theming in our old Objective-C toast class I decided to go the NSNumber route.

I'll send out a new build and will merge once I get final design approval. Thanks!

Comment (rajparad) [reply Wikipedia/Code/EditorViewController.swift]: Thanks for the explanation!!

**Review by tonisevener — APPROVED** [review]
Looks great, thanks @rajparad!

**Comment by rajparad** [comment]
> Looks great, thanks @rajparad!

Thanks for the help!! Merged :) Finally!!!

**Pull request merged**
